internal/exchange/websocket/subscriptions_test.go

```diff
@@ -0,0 +1,305 @@
+	w := NewManager()
+	s := subscription.List{
+		{Key: 42, Channel: "hello3"},
+		{Key: 45, Channel: "hello4"},
+	}
+	err := w.AddSubscriptions(nil, s...)
+	require.NoError(t, err, "AddSubscriptions must not error")
+	assert.ElementsMatch(t, s, w.GetSubscriptions(), "GetSubscriptions should return the correct channel details")
+}
+
+func TestCheckSubscriptions(t *testing.T) {
+	t.Parallel()
+	ws := Manager{}
+	err := ws.checkSubscriptions(nil, nil)
+	assert.ErrorIs(t, err, common.ErrNilPointer, "checkSubscriptions should error correctly on nil w.subscriptions")
+	assert.ErrorContains(t, err, "Websocket.subscriptions", "checkSubscriptions should error giving context correctly on nil w.subscriptions")
+
+	ws.subscriptions = subscription.NewStore()
+	err = ws.checkSubscriptions(nil, nil)
+	assert.NoError(t, err, "checkSubscriptions should not error on a nil list")
+
+	ws.MaxSubscriptionsPerConnection = 1
+
+	err = ws.checkSubscriptions(nil, subscription.List{{}})
+	assert.NoError(t, err, "checkSubscriptions should not error when subscriptions is empty")
+
+	ws.subscriptions = subscription.NewStore()
+	err = ws.checkSubscriptions(nil, subscription.List{{}, {}})
+	assert.ErrorIs(t, err, errSubscriptionsExceedsLimit, "checkSubscriptions should error correctly")
+
+	ws.MaxSubscriptionsPerConnection = 2
+
+	ws.subscriptions = subscription.NewStore()
+	err = ws.subscriptions.Add(&subscription.Subscription{Key: 42, Channel: "test"})
+	require.NoError(t, err, "Add subscription must not error")
+	err = ws.checkSubscriptions(nil, subscription.List{{Key: 42, Channel: "test"}})
+	assert.ErrorIs(t, err, subscription.ErrDuplicate, "checkSubscriptions should error correctly")
+
+	err = ws.checkSubscriptions(nil, subscription.List{{}})
+	assert.NoError(t, err, "checkSubscriptions should not error")
+}
+
+func TestUpdateChannelSubscriptions(t *testing.T) {
+	t.Parallel()
+
+	ws := NewManager()
+	store := subscription.NewStore()
+	err := ws.updateChannelSubscriptions(nil, store, subscription.List{{Channel: "test"}})
+	require.ErrorIs(t, err, common.ErrNilPointer)
+	require.Zero(t, store.Len())
+
+	ws.Subscriber = func(subs subscription.List) error {
+		for _, sub := range subs {
+			if err := store.Add(sub); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	ws.subscriptions = store
+	err = ws.updateChannelSubscriptions(nil, store, subscription.List{{Channel: "test"}})
+	require.NoError(t, err)
+	require.Equal(t, 1, store.Len())
+
+	err = ws.updateChannelSubscriptions(nil, store, subscription.List{})
+	require.ErrorIs(t, err, common.ErrNilPointer)
+
+	ws.Unsubscriber = func(subs subscription.List) error {
+		for _, sub := range subs {
+			if err := store.Remove(sub); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+
+	err = ws.updateChannelSubscriptions(nil, store, subscription.List{})
+	require.NoError(t, err)
+	require.Zero(t, store.Len())
+}
+
+func currySimpleSub(w *Manager) func(subscription.List) error {
```

Comment:
love a good butter chicken

Reply:
FWIW, this recipe is not new.
You've been eating it for quite a while already.